flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryString.java

```diff
@@ -54,21 +62,56 @@ public static BinaryString fromBytes(byte[] bytes) {
 				new MemorySegment[] {MemorySegmentFactory.wrap(bytes)}, 0, bytes.length);
 	}
 
-	public static BinaryString fromBytes(byte[] bytes, int offset, int numBytes) {
-		return new BinaryString(
-				new MemorySegment[] {MemorySegmentFactory.wrap(bytes)}, offset, numBytes);
+	@Override
+	public boolean equals(Object o) {
+		if (o != null && o instanceof BinaryString) {
+			BinaryString other = (BinaryString) o;
+			if (javaObject != null && other.javaObject != null) {
+				return javaObject.equals(other.javaObject);
+			}
+
+			ensureMaterialized();
+			other.ensureMaterialized();
+			return binaryEquals(other);
+		} else {
+			return false;
+		}
+	}
+
+	@Override
+	public int hashCode() {
+		ensureMaterialized();
+		return super.hashCode();
 	}
 
 	@Override
 	public String toString() {
-		byte[] bytes = BinaryStringUtil.allocateReuseBytes(sizeInBytes);
-		SegmentsUtil.copyToBytes(segments, offset, bytes, 0, sizeInBytes);
-		return new String(bytes, 0, sizeInBytes);
+		if (javaObject == null) {
+			byte[] bytes = SegmentsUtil.allocateReuseBytes(sizeInBytes);
+			SegmentsUtil.copyToBytes(segments, offset, bytes, 0, sizeInBytes);
+			javaObject = new String(bytes, 0, sizeInBytes);
+		}
+		return javaObject;
+	}
+
+	public void ensureMaterialized() {
```

Comment:
I think this should belong to `LazyBinaryFormat` ?  And we need to explain what the output of materialization, the bytes or the java object. I don't thins everyone will all in same page if we don't have any comment

Reply:
Their ensureMaterialized method parameters are different. For example, BinaryGeneric needs a TypeSerializer to generate bytes data correctly.